src/Mvc/Mvc.Abstractions/src/ModelBinding/ModelMetadata.cs

```diff
@@ -616,7 +627,8 @@ private void InitializeTypeInformation()
     {
         Debug.Assert(ModelType != null);
 
-        IsComplexType = !TypeDescriptor.GetConverter(ModelType).CanConvertFrom(typeof(string));
+        HasTryParse = ParameterBindingMethodCache.HasTryParseMethod(ModelType);
+        IsComplexType = !HasTryParse && !TypeDescriptor.GetConverter(ModelType).CanConvertFrom(typeof(string));
```

Comment:
@pranavkm, without changing how the `IsComplexType` is set we will validate types bound with `TryParse` as a `ComplexType` instead of `SimpleType` that could cause weird issues like one, below, that I was able to reproduce when using the class `IPEndpoint`

``` csharp 
public ActionResult Get([FromQuery]IPEndPoint iPEndPoint)
```
```
 System.Net.Sockets.SocketException (10045): The attempted operation is not supported for the type of object referenced.
         at System.Net.IPAddress.get_ScopeId()
         at Microsoft.Extensions.Internal.PropertyHelper.CallNullSafePropertyGetter[TDeclaringType,TValue](Func`2 getter, Object target) in C:\git\aspnetcore\src\Shared\PropertyHelper\PropertyHelper.cs:line 396
         at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.DefaultComplexObjectValidationStrategy.Enumerator.GetModel(Object container, ModelMetadata property) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\Validation\DefaultComplexObjectValidationStrategy.cs:line 136
         at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.DefaultComplexObjectValidationStrategy.Enumerator.<>c__DisplayClass13_1.<MoveNext>b__1() in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\Validation\DefaultComplexObjectValidationStrategy.cs:line 118
         at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationEntry.get_Model() in C:\git\aspnetcore\src\Mvc\Mvc.Abstractions\src\ModelBinding\Validation\ValidationEntry.cs:line 86
         at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.VisitChildren(IValidationStrategy strategy) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\Validation\ValidationVisitor.cs:line 455
         at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.VisitComplexType(IValidationStrategy defaultStrategy) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\Validation\ValidationVisitor.cs:line 396
         at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.VisitImplementation(ModelMetadata& metadata, String& key, Object model) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\Validation\ValidationVisitor.cs:line 377
         at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.Visit(ModelMetadata metadata, String key, Object model) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\Validation\ValidationVisitor.cs:line 294
         at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.VisitChildren(IValidationStrategy strategy) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\Validation\ValidationVisitor.cs:line 455
         at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.VisitComplexType(IValidationStrategy defaultStrategy) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\Validation\ValidationVisitor.cs:line 396
         at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.VisitImplementation(ModelMetadata& metadata, String& key, Object model) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\Validation\ValidationVisitor.cs:line 377
         at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.Visit(ModelMetadata metadata, String key, Object model) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\Validation\ValidationVisitor.cs:line 294
         at Microsoft.AspNetCore.Mvc.ModelBinding.Validation.ValidationVisitor.Validate(ModelMetadata metadata, String key, Object model, Boolean alwaysValidateAtTopLevel, Object container) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\Validation\ValidationVisitor.cs:line 205
         at Microsoft.AspNetCore.Mvc.ModelBinding.ObjectModelValidator.Validate(ActionContext actionContext, ValidationStateDictionary validationState, String prefix, Object model, ModelMetadata metadata, Object container) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\ObjectModelValidator.cs:line 110
         at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.EnforceBindRequiredAndValidate(ObjectModelValidator baseObjectValidator, ActionContext actionContext, ParameterDescriptor parameter, ModelMetadata metadata, ModelBindingContext modelBindingContext, ModelBindingResult modelBindingResult, Object container) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\ParameterBinder.cs:line 232
         at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.BindModelAsync(ActionContext actionContext, IModelBinder modelBinder, IValueProvider valueProvider, ParameterDescriptor parameter, ModelMetadata metadata, Object value, Object container) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\ModelBinding\ParameterBinder.cs:line 182
         at Microsoft.AspNetCore.Mvc.Controllers.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext() in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\Controllers\ControllerBinderDelegateProvider.cs:line 92
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\Infrastructure\ControllerActionInvoker.cs:line 484
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\Infrastructure\ResourceInvoker.cs:line 250
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\Infrastructure\ResourceInvoker.cs:line 124
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker) in C:\git\aspnetcore\src\Mvc\Mvc.Core\src\Infrastructure\ResourceInvoker.cs:line 130
         at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
         at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
         at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)


```

One approach that I thought about was trying to set the `IsComplexType` when the type has a `TryParse` only when the new `BinderProvider` is available but I don't think it make totally sense.

Do you have any suggestions on how can I avoid this change (maybe is ok validate them as complex types)?

Reply:
This seems fine-ish. We're treating these new models as being deserialized from a single string, so it feels right. That said this needs to be affected by the presence of the binder - if it has been removed and does not have a type converter, it's going to fall back to complex type binding but validation would treat it differently.

Reply:
Yeah! That was my concern since the beginning. Let me see what I can do here in the PR and we can update the API proposal and get it back for the review.